### PR TITLE
Add FixedSizeBinaryArray::try_from_sparse_iter_with_size

### LINKED ISSUE #1390: `FixedSizeBinaryArray::try_from_sparse_iter` failed when given all Nones
**Describe the bug**

try_from_sparse_iter expects at least one item to have Some to determine the size. This is a bad assumption as it may recieve all Nones. 

```
InvalidArgumentError("column types must match schema types, expected FixedSizeBinary(32) but found FixedSizeBinary(0) at column index 0"
```
**To Reproduce**

```
        let none_option: Option<[u8; 32]> = None;
        let item = arrow::array::FixedSizeBinaryArray::try_from_sparse_iter(
            vec![none_option, none_option, none_option].into_iter(),
        )
        .unwrap();
       let rb = arrow::record_batch::RecordBatch::try_new(
            arrow_schema.clone(),
            vec![std::sync::Arc::new(item)],
        )
        .unwrap();

```

**Expected behavior**

A record batch with a colum of all nulls.

**Additional context**

I would suggest fixing it by changing the behaviour of defaulting in **array_binary.rs:560**

```
let size = size.unwrap_or(0);
```

or changing the schema validation to handle a 0 size if the array is empty.

### PR DESCRIPTION
**Pull request by maxburke**
Closes #1390

# Rationale for this change
 
FixedSizeBinaryArray::try_from_sparse_iter generates incorrect results when the provided iterator only produces None values.

# What changes are included in this PR?

This change also implements the comparison kernel for FixedSizeBinary types.

# Are there any user-facing changes?

This change adds a new function, FixedSizeBinaryArray::try_from_iter_with_size, which takes a default size parameter, and markes the existing `try_from_sparse_iter` as deprecated.

### CODE REVIEW

File: arrow-array/src/array/fixed_size_binary_array.rs [hunk 1]
```diff
@@ -129,17 +129,29 @@ impl FixedSizeBinaryArray {
...
+    where
+        T: Iterator<Item = Option<U>>,
+        U: AsRef<[u8]>,
+    {
+        Self::try_from_sparse_iter_with_size(iter, None)
+    }
+
+    pub fn try_from_sparse_iter_with_size<T, U>(
+        mut iter: T,
+        asserted_size: Option<i32>,
```
Comment (tustvold) [inline arrow-array/src/array/fixed_size_binary_array.rs:142]: I don't follow the need for a separate asserted_size and detected_size. Looking at the logic I think we could just have `mut size: Option<i32>` and everything would work correctly?

File: arrow/src/compute/kernels/comparison.rs [hunk 2]
```diff
@@ -2270,6 +2270,18 @@ macro_rules! typed_compares {
                 as_largestring_array($RIGHT),
                 $OP,
             ),
+            (DataType::FixedSizeBinary(_), DataType::FixedSizeBinary(_)) => {
```
Comment (tustvold) [inline arrow/src/compute/kernels/comparison.rs:2273]: This change does not appear to be related?

File: arrow-array/src/array/fixed_size_binary_array.rs [hunk 3]
```diff
@@ -129,17 +129,29 @@ impl FixedSizeBinaryArray {
...
-    pub fn try_from_sparse_iter<T, U>(mut iter: T) -> Result<Self, ArrowError>
+    pub fn try_from_sparse_iter<T, U>(iter: T) -> Result<Self, ArrowError>
+    where
+        T: Iterator<Item = Option<U>>,
+        U: AsRef<[u8]>,
+    {
+        Self::try_from_sparse_iter_with_size(iter, None)
+    }
+
+    pub fn try_from_sparse_iter_with_size<T, U>(
```
Comment (tustvold) [inline arrow-array/src/array/fixed_size_binary_array.rs:140]: I think a doc comment similar to the one above would be good, especially to explain what the size argument is

Comment (tustvold) [inline arrow-array/src/array/fixed_size_binary_array.rs:140]:
```suggestion
    pub fn try_from_sparse_iter_with_size<T, U>(iter: T, size: i32) -> Result<Self, ArrowError> {
        try_from_sparse_iter_with_size_opt::<T, U>(iter, Some(size))
    }
    
    fn try_from_sparse_iter_with_size_opt<T, U>(
```

I think this makes for better ergonomics

**Review by tustvold — COMMENTED** [review]
Left some comments

File: arrow-array/src/array/fixed_size_binary_array.rs (hunk 1 above)
Comment (maxburke) [reply arrow-array/src/array/fixed_size_binary_array.rs:142]: I think if we were to just have a single size option we would need to have separate implementations of `try_from_sparse_iter` and `try_from_sparse_iter_with_size`.

(IMO, I think the former is problematic and should be deprecated / removed)

File: arrow/src/compute/kernels/comparison.rs (hunk 2 above)
Comment (maxburke) [reply arrow/src/compute/kernels/comparison.rs:2273]: Yeah, it's not, but it was something we were hoping to upstream and I think it accidentally fell into the branch. Would you like me to update this PR to mention this change?

Comment (tustvold) [reply arrow/src/compute/kernels/comparison.rs:2273]: I think either we need tests of it, or we need to roll it back

File: arrow-array/src/array/fixed_size_binary_array.rs (hunk 1 above)
Comment (tustvold) [reply arrow-array/src/array/fixed_size_binary_array.rs:142]: I don't see why you need to distinguish between the two different sizes in the implementation, it doesn't change if the size is determined by being passed in or from the first non-null element. They must all be equal regardless?

**Comment by alamb** [comment]
If we can get this PR merged in the next 4-5 hours it will be included in arrow 27.0.0.

File: arrow/src/compute/kernels/comparison.rs (hunk 2 above)
Comment (maxburke) [reply arrow/src/compute/kernels/comparison.rs:2273]: Tests added.

File: arrow-array/src/array/fixed_size_binary_array.rs (hunk 3 above)
Comment (maxburke) [reply arrow-array/src/array/fixed_size_binary_array.rs]: Done.

Comment (maxburke) [reply arrow-array/src/array/fixed_size_binary_array.rs]: I made the parameter Optional instead

File: arrow-array/src/array/fixed_size_binary_array.rs (hunk 1 above)
Comment (maxburke) [reply arrow-array/src/array/fixed_size_binary_array.rs]: I separated the instances of `try_from_sparse_iter` and `try_from_sparse_iter_with_size` because I think the corner cases of making the two implementations match up wasn't worth the complexity.

**Comment by maxburke** [comment]
> If we can get this PR merged in the next 4-5 hours it will be included in arrow 27.0.0.

rats

**Comment by alamb** [comment]
> If we can get this PR merged in the next 4-5 hours it will be included in arrow 27.0.0.

😢

**Review by tustvold — APPROVED** [review]

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
But there are a few places using the deprecated function, see CI.

```
error: use of deprecated associated function `arrow_array::FixedSizeBinaryArray::try_from_sparse_iter`: This function will fail if the iterator produces only None values; prefer `try_from_sparse_iter_with_size`
   --> arrow/src/array/ffi.rs:215:43
    |
215 |         let array = FixedSizeBinaryArray::try_from_sparse_iter(values.into_iter())?;
    |                                           ^^^^^^^^^^^^^^^^^^^^
    |
    = note: `-D deprecated` implied by `-D warnings`
```

**Comment by maxburke** [comment]
> But there are a few places using the deprecated function, see CI.
> 
> ```
> error: use of deprecated associated function `arrow_array::FixedSizeBinaryArray::try_from_sparse_iter`: This function will fail if the iterator produces only None values; prefer `try_from_sparse_iter_with_size`
>    --> arrow/src/array/ffi.rs:215:43
>     |
> 215 |         let array = FixedSizeBinaryArray::try_from_sparse_iter(values.into_iter())?;
>     |                                           ^^^^^^^^^^^^^^^^^^^^
>     |
>     = note: `-D deprecated` implied by `-D warnings`
> ```

Yup, just saw the build failure and submitted a fix. I wasn't testing locally with the `ffi` feature enabled.

**Comment by viirya** [comment]
There are some clippy errors that are not related to this change. Submitted #3096 to fix that.

**Comment by tustvold** [comment]
@maxburke could you possibly merge master into this branch, so the CI goes green. I don't seem to have permissions to do this

**Comment by maxburke** [comment]
> @maxburke could you possibly merge master into this branch, so the CI goes green. I don't seem to have permissions to do this, there is a checkbox to allow maintainers the ability to edit PRed branches

done!

**Comment by viirya** [comment]
Thanks @maxburke

**Comment by alamb** [comment]
🎉
